File: Src/DetailedSamples/Visual Basic .NET/Multiple/Xceed Windows Explorer/Utils/ListView/ListViewSortComparer.vb
'
 '* Xceed Zip for .NET - Xceed Windows Explorer sample application
 '* Copyright (c) 2006 - Xceed Software Inc.
 '*
'* [ListViewSortComparer.vb]
 '*
 '* Custom SortComparer that mimic the MS Windows Explorer sorting.
 '*
 '* This file is part of Xceed Zip for .NET. The source code in this file
 '* is only intended as a supplement to the documentation, and is provided
 '* "as is", without warranty of any kind, either expressed or implied.
 '


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Windows.Forms
Imports Xceed.FileSystem
Imports Xceed.FileSystem.Samples.Utils.TreeView
Imports Xceed.Zip

Namespace Xceed.FileSystem.Samples.Utils.ListView
  Public Class ListViewSortComparer : Implements IComparer
    #Region "CONSTRUCTORS"

    Public Sub New(ByVal colIndex As Integer)
      m_columnIndex = colIndex
      m_sortOrder = SortOrder.None
    End Sub

    Public Sub New(ByVal colIndex As Integer, ByVal order As SortOrder)
      m_columnIndex = colIndex
      m_sortOrder = order
    End Sub

#End Region     ' CONSTRUCTORS

    #Region "PUBLIC PROPERTIES"

    ''' <summary>
    ''' Gets or sets the column index on which to sort the list.
    ''' </summary>
    Public Property ColumnIndex() As Integer
      Get
        Return m_columnIndex
      End Get
      Set
        m_columnIndex = Value
      End Set
    End Property

    ''' <summary>
    ''' Gets or sets the sort order.
    ''' </summary>
    Public Property SortOrder() As SortOrder
      Get
        Return m_sortOrder
      End Get
      Set
        m_sortOrder = Value
      End Set
    End Property

    #End Region ' PUBLIC PROPERTIES

    #Region "IComparer Members"

    Public Function Compare(ByVal x As Object, ByVal y As Object) As Integer Implements IComparer.Compare
      ' The objects return here contains a FileSystemItem which will be used to compare.
      Dim xItem As AbstractListViewItem = CType(IIf(TypeOf x Is AbstractListViewItem, x, Nothing), AbstractListViewItem)
      Dim yItem As AbstractListViewItem = CType(IIf(TypeOf y Is AbstractListViewItem, y, Nothing), AbstractListViewItem)

      Select Case m_columnIndex
        Case 0 ' Name
          Return Me.CompareName(xItem, yItem)

        Case 1 ' Size
          Return Me.CompareSize(xItem, yItem)

        Case 2 ' Compressed Size
          Return Me.CompareCompressedSize(xItem, yItem)

        Case 3 ' Type
          Return Me.CompareType(xItem, yItem)

        Case 4 ' Modification date
          Return Me.CompareModificationDate(xItem, yItem)

        Case Else
          Return 0
      End Select
    End Function

    #End Region ' IComparer Members

    #Region "PRIVATE METHODS"

    ''' <summary>
    ''' Compare the compressed size of 2 AbstractListViewItem.
    ''' </summary>
    ''' <param name="xItem"></param>
    ''' <param name="yItem"></param>
    ''' <returns>Returns -1 if xItem is less than yItem, 1 if xItem is
    ''' greater than yItem and 0 if they are equal. Folders are compared 
    ''' using the display text. The values will be reverted if the sort order
    ''' is descending.</returns>
    Private Function CompareCompressedSize(ByVal xItem As AbstractListViewItem, ByVal yItem As AbstractListViewItem) As Integer
      ' Folders will have a size of -1.
      Dim xSize As Long = -1
      Dim ySize As Long = -1

      ' Folder should be sorted by name.
      Dim xFolder As AbstractFolder = CType(IIf(TypeOf xItem.FileSystemItem Is ZippedFolder, xItem.FileSystemItem, Nothing), ZippedFolder)
      Dim yFolder As AbstractFolder = CType(IIf(TypeOf yItem.FileSystemItem Is ZippedFolder, yItem.FileSystemItem, Nothing), ZippedFolder)

      If Not xFolder Is Nothing AndAlso Not yFolder Is Nothing Then
        Return Me.GetCompareResult(String.Compare(xItem.Text, yItem.Text))
      End If

      Dim xFile As AbstractFile = CType(IIf(TypeOf xItem.FileSystemItem Is ZippedFile, xItem.FileSystemItem, Nothing), ZippedFile)
      Dim yFile As AbstractFile = CType(IIf(TypeOf yItem.FileSystemItem Is ZippedFile, yItem.FileSystemItem, Nothing), ZippedFile)

      If Not xFile Is Nothing Then
        xSize = (CType(xFile, ZippedFile)).CompressedSize
      End If

      If Not yFile Is Nothing Then
        ySize = (CType(yFile, ZippedFile)).CompressedSize
      End If

      If xSize > ySize Then
        Return Me.GetCompareResult(1)
      Else If xSize < ySize Then
        Return Me.GetCompareResult(-1)
      Else ' Equal size, we sort by name.
        Return Me.GetCompareResult(String.Compare(yItem.Text, xItem.Text))
      End If
    End Function

    Private Function CompareModificationDate(ByVal xItem As AbstractListViewItem, ByVal yItem As AbstractListViewItem) As Integer
      ' If x is a folder and y is not, x shoud come first.
      If (Not CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFolder, xItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) AndAlso (CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFolder, yItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) Then
        Return Me.GetCompareResult(-1)
      End If

      ' If y is a folder and x is not, y shoud come first.
      If (CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFolder, xItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) AndAlso (Not CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFolder, yItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) Then
        Return Me.GetCompareResult(1)
      End If

      Dim xDateTime As DateTime = DateTime.MinValue
      Dim yDateTime As DateTime = DateTime.MinValue

      If xItem.FileSystemItem.HasLastWriteDateTime Then
        xDateTime = xItem.FileSystemItem.LastWriteDateTime
      End If

      If yItem.FileSystemItem.HasLastWriteDateTime Then
        yDateTime = yItem.FileSystemItem.LastWriteDateTime
      End If

      If xDateTime > yDateTime Then
        Return Me.GetCompareResult(1)
      ElseIf xDateTime < yDateTime Then
        Return Me.GetCompareResult(-1)
      Else ' Equal size, we sort by name.
        Return Me.GetCompareResult(String.Compare(yItem.Text, xItem.Text))
      End If
    End Function

    Private Function CompareName(ByVal xItem As AbstractListViewItem, ByVal yItem As AbstractListViewItem) As Integer
      ' If x is a folder and y is not, x shoud come first.
      If (Not CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFolder, xItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) AndAlso (CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFolder, yItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) Then
        Return Me.GetCompareResult(-1)
      End If

      ' If y is a folder and x is not, y shoud come first.
      If (CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFolder, xItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) AndAlso (Not CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFolder, yItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) Then
        Return Me.GetCompareResult(1)
      End If

      ' We are dealing with the same type of item, we just compare the 2 strings.
      Return Me.GetCompareResult(String.Compare(xItem.Text, yItem.Text))
    End Function

    Private Function CompareSize(ByVal xItem As AbstractListViewItem, ByVal yItem As AbstractListViewItem) As Integer
      ' Folders will have a size of -1.
      Dim xSize As Long = -1
      Dim ySize As Long = -1

      ' Folder should be sorted by name.
      Dim xFolder As AbstractFolder = CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFolder, xItem.FileSystemItem, Nothing), AbstractFolder)
      Dim yFolder As AbstractFolder = CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFolder, yItem.FileSystemItem, Nothing), AbstractFolder)

      If Not xFolder Is Nothing AndAlso Not yFolder Is Nothing Then
        Return Me.GetCompareResult(String.Compare(xItem.Text, yItem.Text))
      End If

      Dim xFile As AbstractFile = CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFile, xItem.FileSystemItem, Nothing), AbstractFile)
      Dim yFile As AbstractFile = CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFile, yItem.FileSystemItem, Nothing), AbstractFile)

      If Not xFile Is Nothing Then
        xSize = xFile.Size
      End If

      If Not yFile Is Nothing Then
        ySize = yFile.Size
      End If

      If xSize > ySize Then
        Return Me.GetCompareResult(1)
      Else If xSize < ySize Then
        Return Me.GetCompareResult(-1)
      Else ' Equal size, we sort by name.
        Return Me.GetCompareResult(String.Compare(yItem.Text, xItem.Text))
      End If
    End Function

    Private Function CompareType(ByVal xItem As AbstractListViewItem, ByVal yItem As AbstractListViewItem) As Integer
      ' If x is a folder and y is not, x shoud come first.
      If (Not CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFolder, xItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) AndAlso (CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFolder, yItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) Then
        Return Me.GetCompareResult(-1)
      End If

      ' If y is a folder and x is not, y shoud come first.
      If (CType(IIf(TypeOf xItem.FileSystemItem Is AbstractFolder, xItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) AndAlso (Not CType(IIf(TypeOf yItem.FileSystemItem Is AbstractFolder, yItem.FileSystemItem, Nothing), AbstractFolder) Is Nothing) Then
        Return Me.GetCompareResult(1)
      End If

      Dim xType As String = xItem.FileSystemItem.GetType().ToString()
      Dim yType As String = yItem.FileSystemItem.GetType().ToString()

      If xType <> yType Then
        Return Me.GetCompareResult(String.Compare(yType, xType))
      Else
        ' Same type, we sort by name.
        Return Me.GetCompareResult(String.Compare(yItem.Text, xItem.Text))
      End If
    End Function

    Private Function GetCompareResult(ByVal value As Integer) As Integer
      Select Case m_sortOrder
        Case SortOrder.Ascending
          Return value

        Case SortOrder.Descending
          Return -value

        Case Else
          Return 0
      End Select
    End Function

    #End Region ' PRIVATE METHODS

    #Region "PRIVATE FIELDS"

    Private m_sortOrder As SortOrder
    Private m_columnIndex As Integer = -1

    #End Region ' PRIVATE FIELDS
  End Class
End Namespace

File: Src/DetailedSamples/CSharp/Multiple/Xceed Windows Explorer/Utils/ListView/ListViewSortComparer.cs
/*
 * Xceed Zip for .NET - Xceed Windows Explorer sample application
 * Copyright (c) 2006 - Xceed Software Inc.
 * 
 * [ListViewSortComparer.cs]
 * 
 * Custom SortComparer that mimic the MS Windows Explorer sorting.
 * 
 * This file is part of Xceed Zip for .NET. The source code in this file 
 * is only intended as a supplement to the documentation, and is provided 
 * "as is", without warranty of any kind, either expressed or implied.
 */

using System;
using System.Collections;
using System.Windows.Forms;
using Xceed.FileSystem;
using Xceed.FileSystem.Samples.Utils.TreeView;
using Xceed.Zip;

namespace Xceed.FileSystem.Samples.Utils.ListView
{
	public class ListViewSortComparer : IComparer
	{
    #region CONSTRUCTORS

		public ListViewSortComparer( int columnIndex )
      : this( columnIndex, SortOrder.None )
    {
    }

    public ListViewSortComparer( int columnIndex, SortOrder sortOrder )
    {
      m_columnIndex = columnIndex;
      m_sortOrder = sortOrder;
    }

    #endregion CONSTRUCTORS

    #region PUBLIC PROPERTIES

    /// <summary>
    /// Gets or sets the column index on which to sort the list.
    /// </summary>
    public int ColumnIndex
    {
      get{ return m_columnIndex; }
      set{ m_columnIndex = value; }
    }
    
    /// <summary>
    /// Gets or sets the sort order.
    /// </summary>
    public SortOrder SortOrder
    {
      get{ return m_sortOrder; }
      set{ m_sortOrder = value; }
    }

    #endregion PUBLIC PROPERTIES

    #region IComparer Members

    public int Compare(object x, object y)
    {
      // The objects return here contains a FileSystemItem which will be used to compare.
      AbstractListViewItem xItem = x as AbstractListViewItem;
      AbstractListViewItem yItem = y as AbstractListViewItem;

      switch( m_columnIndex )
      {
        case 0: // Name
          return this.CompareName( xItem, yItem );

        case 1: // Size
          return this.CompareSize( xItem, yItem );

        case 2: // Compressed Size
          return this.CompareCompressedSize( xItem, yItem );

        case 3: // Type
          return this.CompareType( xItem, yItem );

        case 4: // Modification date
          return this.CompareModificationDate( xItem, yItem );          

        default:
          return 0;
      }
    }

    #endregion IComparer Members

    #region PRIVATE METHODS

    /// <summary>
    /// Compare the compressed size of 2 AbstractListViewItem.
    /// </summary>
    /// <param name="xItem"></param>
    /// <param name="yItem"></param>
    /// <returns>Returns -1 if xItem is less than yItem, 1 if xItem is
    /// greater than yItem and 0 if they are equal. Folders are compared 
    /// using the display text. The values will be reverted if the sort order
    /// is descending.</returns>
    private int CompareCompressedSize( AbstractListViewItem xItem, AbstractListViewItem yItem )
    {
      // Folders will have a size of -1.
      long xSize = -1;
      long ySize = -1;

      // Folder should be sorted by name.
      AbstractFolder xFolder = xItem.FileSystemItem as ZippedFolder;
      AbstractFolder yFolder = yItem.FileSystemItem as ZippedFolder;

      if( xFolder != null && yFolder != null )
      {
        return this.GetCompareResult( string.Compare( xItem.Text, yItem.Text ) );
      }

      AbstractFile xFile = xItem.FileSystemItem as ZippedFile;
      AbstractFile yFile = yItem.FileSystemItem as ZippedFile;

      if( xFile != null )
        xSize = ( ( ZippedFile )xFile ).CompressedSize;

      if( yFile != null )
        ySize = ( ( ZippedFile )yFile ).CompressedSize;

      if( xSize > ySize )
      {
        return this.GetCompareResult( 1 );
      }
      else if( xSize < ySize )
      {
        return this.GetCompareResult( -1 );
      }
      else // Equal size, we sort by name.
      {
        return this.GetCompareResult( string.Compare( yItem.Text, xItem.Text ) );
      }
    }

    private int CompareModificationDate( AbstractListViewItem xItem, AbstractListViewItem yItem )
    {
      // If x is a folder and y is not, x shoud come first.
      if(  ( xItem.FileSystemItem as AbstractFolder != null ) 
        && ( yItem.FileSystemItem as AbstractFolder == null ) )
      {
        return this.GetCompareResult( -1 );
      }

      // If y is a folder and x is not, y shoud come first.
      if(  ( xItem.FileSystemItem as AbstractFolder == null ) 
        && ( yItem.FileSystemItem as AbstractFolder != null ) )
      {
        return this.GetCompareResult( 1 );
      }

      DateTime xDateTime = DateTime.MinValue;
      DateTime yDateTime = DateTime.MinValue;

      if( xItem.FileSystemItem.HasLastWriteDateTime )
        xDateTime = xItem.FileSystemItem.LastWriteDateTime;

      if( yItem.FileSystemItem.HasLastWriteDateTime )
        yDateTime = yItem.FileSystemItem.LastWriteDateTime;

      if( xDateTime > yDateTime )
      {
        return this.GetCompareResult( 1 );
      }
      else if( xDateTime < yDateTime )
      {
        return this.GetCompareResult( -1 );
      }
      else // Equal size, we sort by name.
      {
        return this.GetCompareResult( string.Compare( yItem.Text, xItem.Text ) );
      }
    }

    private int CompareName( AbstractListViewItem xItem, AbstractListViewItem yItem )
    {
      // If x is a folder and y is not, x shoud come first.
      if(  ( xItem.FileSystemItem as AbstractFolder != null ) 
        && ( yItem.FileSystemItem as AbstractFolder == null ) )
      {
        return this.GetCompareResult( -1 );
      }

      // If y is a folder and x is not, y shoud come first.
      if(  ( xItem.FileSystemItem as AbstractFolder == null ) 
        && ( yItem.FileSystemItem as AbstractFolder != null ) )
      {
        return this.GetCompareResult( 1 );
      }

      // We are dealing with the same type of item, we just compare the 2 strings.
      return this.GetCompareResult( string.Compare( xItem.Text, yItem.Text ) );
    }

    private int CompareSize( AbstractListViewItem xItem, AbstractListViewItem yItem )
    {
      // Folders will have a size of -1.
      long xSize = -1;
      long ySize = -1;

      // Folder should be sorted by name.
      AbstractFolder xFolder = xItem.FileSystemItem as AbstractFolder;
      AbstractFolder yFolder = yItem.FileSystemItem as AbstractFolder;

      if( xFolder != null && yFolder != null )
      {
        return this.GetCompareResult( string.Compare( xItem.Text, yItem.Text ) );
      }

      AbstractFile xFile = xItem.FileSystemItem as AbstractFile;
      AbstractFile yFile = yItem.FileSystemItem as AbstractFile;

      if( xFile != null )
        xSize = xFile.Size;

      if( yFile != null )
        ySize = yFile.Size;

      if( xSize > ySize )
      {
        return this.GetCompareResult( 1 );
      }
      else if( xSize < ySize )
      {
        return this.GetCompareResult( -1 );
      }
      else // Equal size, we sort by name.
      {
        return this.GetCompareResult( string.Compare( yItem.Text, xItem.Text ) );
      }
    }

    private int CompareType( AbstractListViewItem xItem, AbstractListViewItem yItem )
    {
      // If x is a folder and y is not, x shoud come first.
      if(  ( xItem.FileSystemItem as AbstractFolder != null ) 
        && ( yItem.FileSystemItem as AbstractFolder == null ) )
      {
        return this.GetCompareResult( -1 );
      }

      // If y is a folder and x is not, y shoud come first.
      if(  ( xItem.FileSystemItem as AbstractFolder == null ) 
        && ( yItem.FileSystemItem as AbstractFolder != null ) )
      {
        return this.GetCompareResult( 1 );
      }

      string xType = xItem.FileSystemItem.GetType().ToString();
      string yType = yItem.FileSystemItem.GetType().ToString();

      if( xType != yType )
      {
        return this.GetCompareResult( string.Compare( yType, xType ) );
      }
      else
      {
        // Same type, we sort by name.
        return this.GetCompareResult( string.Compare( yItem.Text, xItem.Text ) );
      }
    }

    private int GetCompareResult( int value )
    {
      switch( m_sortOrder )
      {
        case SortOrder.Ascending:
          return value;

        case SortOrder.Descending:
          return -value;

        default:
          return 0;
      }
    }

    #endregion PRIVATE METHODS

    #region PRIVATE FIELDS

    private SortOrder m_sortOrder;
    private int m_columnIndex = -1;

    #endregion PRIVATE FIELDS
  }
}
